draft-ietf-masque-connect-udp.md

```diff
@@ -45,119 +55,222 @@ document are to be interpreted as described in BCP 14 {{!RFC2119}} {{!RFC8174}}
+"proxy_host" and "proxy_port" are the configured host and port of the proxy
+respectively. Proxy deployments SHOULD use the default template to facilitate
+interoperability with such clients.
+
+
+# HTTP Exchanges
+
+This document defines the "masque-udp" HTTP Upgrade Token. "masque-udp" uses
+the Capsule Protocol as defined in {{HTTP-DGRAM}}.
+
+A "masque-udp" request requests that the recipient establish a tunnel over a
+single HTTP stream to the destination target server identified by the
+"target_host" and "target_port" variables of the URI template (see
+{{client-config}}). If the request is successful, the proxy commits to
+converting received HTTP Datagrams into UDP packets and vice versa until the
+tunnel is closed. Tunnels are commonly used to create an end-to-end virtual
+connection, which can then be secured using QUIC {{!QUIC=RFC9000}} or another
+protocol running over UDP.
+
+When sending its UDP proxying request, the client SHALL perform URI template
+expansion to determine the path and query of its request. target_host supports
+using DNS names, IPv6 literals and IPv4 literals. Note that this URI template
+expansion requires using pct-encoding, so for example if the target_host is
+"2001:db8::42", it will be encoded in the URI as "2001%3Adb8%3A%3A42".
+
+A payload within a UDP proxying request message has no defined semantics; a UDP
+proxying request with a non-empty payload is malformed.
+
+Responses to UDP proxying requests are not cacheable.
+
+## Proxy Handling
+
+Upon receiving a UDP proxying request, the recipient proxy extracts the
+"target_host" and "target_port" variables from the URI it has reconstructed
+from the request headers, and establishes a tunnel by directly opening a UDP
+socket to the requested target.
+
+Unlike TCP, UDP is connection-less. The proxy that opens the UDP socket has no
+way of knowing whether the destination is reachable. Therefore it needs to
+respond to the request without waiting for a TCP SYN-ACK.
+
+Proxies can use connected UDP sockets if their operating system supports them,
+as that allows the proxy to rely on the kernel to only send it UDP packets that
+match the correct 5-tuple. If the proxy uses a non-connected socket, it MUST
+validate the IP source address and UDP source port on received packets to
+ensure they match the client's request. Packets that do not match MUST be
+discarded by the proxy.
+
+The lifetime of the socket is tied to the request stream. The proxy MUST keep
+the socket open while the request stream is open. Proxies MAY choose to close
+sockets due to a period of inactivity, but they MUST close the request stream
+before closing the socket. If a proxy is notified by its operating system that
+its socket is no longer usable, it MUST close the request stream.
+
+A successfull response (as defined in {{resp1}} and {{resp23}}) indicates that
+the proxy has opened a socket to the requested target and is willing to proxy
```

Comment:
I think we also need to mention that if hostname resolution is required, the proxy MUST do that before responding, and if there is an error SHOULD send it back via a proxy-status header.

Reply:
Good point, done.